docs/architecture/adr-005-subscription.md

```diff
@@ -0,0 +1,225 @@
+
+For fuller context around this issue: see [\#4642](https://github.com/cosmos/cosmos-sdk/issues/4642)
+
+## Decision
+
+Create Subscription module in SDK so that users can subscribe to various subscription services. Subscription collectors will send messages to collect subscription payments from users who have due subscriptions. Subscriptions are paid out directly from a user account. If the user does not have enough funds to pay for the subscription is inactivated. Users can define maximum limits on how many periods the subscription is valid for; if this limit elapses before a subscription renewal, the subscription is invalidated.
+
+## Status
+
+#### Proposed
+
+## Implementation Changes
+
+Introduce the following messages to `x/subscription` module:
+
+```go
+// CreateSubscriptionMsg allows Service Provider to create new subscription service.
+type CreateSubscriptionMsg struct {
+    Name      string         // unique, human-readable name for subscription service
+    Amount    sdk.Coins      // amount to be collected for each subscription period
+    Period    time.time      // duration of subscription period
+    Collector sdk.AccAddress // address that will collect subscription payments
+}
+```
+
+```go
+// SubscribeMsg allows subscriber account to subscribe to provided service
+// If subscription to the service from user already exists, this msg is treated as renewal.
+// If msg.Limit != -1, then subscription.Limit += msg.Limit. Else, subscription.Limit = -1 (unlimited)
+type SubscribeMsg struct {
+    Name       string         // name of service to subscribe to
+    Subscriber sdk.AccAddress // address of subscriber
+    Limit      int64          // Maximum number of periods that subscription remains active. Limit = -1 implies no limit
+}
+```
+
+```go
+// UnsubscribeMsg allows subscriber to inactivate an active subscription
+type UnsubscribeMsg struct {
+    Name       string         // name of service to unsubscribe to
+    Subscriber sdk.AccAddress // address of subscriber
+}
+```
+
+```go
+// CollectMsg allows a Collector for a service to collect payments on due subscriptions that are processed off a FIFO queue.
+type CollectMsg struct {
+    Name      string         // name of subscription service to collect payments from
+    Collector sdk.AccAddress // address that will collect payments
+    Limit     int64          // maximum number of items to process in FIFO duequeue. If Limit = -1, try to process all due subscriptions
+}
+```
+
+Create a new store with the following key-values:
+
+`Address => []SubscriptionID // List of active/inactive subscriptions owned by the users`
+
+`Metadata:{Name} => SubscriptionMetaData`
+
+`DueQueue:{Name} => []SubscriptionID // FIFO queue of subscriptions. CONTRACT: All due subscriptions exist before all undue subscriptions. All subscriptions in DueQueue are active`
```

Comment:
Idea for single `Period` only: 

As the tree is deterministic and we don't need to keep track of two lists, I think an alternative solution would be deleting the subscription from the store and then adding it again. That'd give us a list of subscriptions sorted by due time.

Reply:
The second list for `Address => []SubscriptionID` is so that users can easily query fo all subscriptions both active and inactive that their account has subscribed to. Doesn't have anything to do with collection, it is for help with querying